### ISSUE TITLE
Possible while_immutable_condition false positive

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust-clippy#6689]
This uses a Rust idiom:

```rust
fn foo(mut n: u64) -> u32 {
    let mut count = 0;
    while { n >>= 1; n != 0 } {
        count += 1;
    }
    count
}

fn main() {
    println!("{}", foo(1023));
    println!("{}", foo(1024));
}
```

Clippy with rustc 1.51.0-nightly gives:

```rust
error: variables in the condition are not mutated in the loop body
 --> test.rs:3:11
  |
3 |     while { n >>= 1; n != 0 } {
  |           ^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[deny(clippy::while_immutable_condition)]` on by default
  = note: this may lead to an infinite or to a never running loop
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#while_immutable_condition

error: aborting due to previous error
```

**Comment by xfix** [comment]
Personally I would use `loop` instead. Using `while` like this feels needlessly clever to me.

```rust
fn foo(mut n: u64) -> u32 {
    let mut count = 0;
    loop {
        n >>= 1;
        if n == 0 {
            break;
        }
        count += 1;
    }
    count
}

fn main() {
    println!("{}", foo(1023));
    println!("{}", foo(1024));
}
```
